### PR TITLE
feat: Separated UI from core
    
    
    
    BREAKING CHANGE:  _______

### PR DESCRIPTION
**Pull request by kujtimprenkuSQA**
# Description
The main goal of this PR is to separate the **Modal UI** into a separate package and move it away from the `core` package.
The idea is to give users/developers more flexibility, now they will have the option to use our Modal or they can implement their own UI.
The modal now will be a custom web component which is developed using [Stencil](https://stenciljs.com/docs/introduction).

This PR adds three additional projects under `/packages:`
- **selector-ui** (the stencil component from which we create the react and angular wrappers around)
- **selector-ui-react** ( creates `<WalletSelectorModal />`)
- **selector-ui-angular** (creates `<wallet-selector-modal />`)

These are the main things on the PR:
- Removed old modal.
- Created Modal UI using Stencil.
- Created 3 additional packages/projects mentioned above.
- Updated the react and angular example to use the new modal component(s).

Closes # (issue)

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

# Type of change. This type of change is the main reason for the PR.

- [ ] FIX - a PR of this type patches a bug.
- [x] FEATURE - a PR of this type introduces a new feature.
- [ ] BUILD - a PR of this type introduces build changes.
- [ ] CI - a PR of this type introduces CI changes.
- [ ] DOCS - a PR of this type introduces DOCS improvement.
- [ ] STYLE - a PR of this type introduces style changes.
- [ ] REFACTOR - a PR of this type introduces refactoring.
- [ ] PERFORMANCE - a PR of this type introduces performance changes.
- [ ] TEST - a PR of this type adds more tests.
- [ ] CHORE - a PR introduces other changes than the specified above.

# Breaking changes

- [x] BREAKING CHANGE - SPECIFY: _______
- [ ] NO BREAKING CHANGE - this PR doesn't contain any breaking changes and it's backwards compatible

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -1,3 +1,6 @@
 node_modules
 examples/react/out/
 examples/react/dist/
+packages/selector-ui-angular/src/generated/
+packages/selector-ui-react/src/generated/
+packages/selector-ui-svelte/src/generated/
```
Comment (lewis-sqa) [inline .eslintignore:6]: Could we remove `svelte` for now until we ever get demand for it?

File: README.md [hunk 2]
```diff
@@ -30,20 +30,10 @@ Next, you'll need to install the wallets you want to support:
...
-npm install \
-  @near-wallet-selector/near-wallet \
-  @near-wallet-selector/sender \
-  @near-wallet-selector/math-wallet \
-  @near-wallet-selector/ledger \
-  @near-wallet-selector/wallet-connect
+npm install @near-wallet-selector/{near-wallet,sender,math-wallet,ledger,wallet-connect}
 ```
 
 Then them in your dApp:
```
Comment (lewis-sqa) [inline README.md:39]: We'll want to update the docs to show how they would setup the modal now

File: examples/angular/src/app/components/content/content.component.ts [hunk 3]
```diff
@@ -59,7 +59,7 @@ export class ContentComponent implements OnInit, OnDestroy {
   }
 
   signIn() {
-    this.selector.show();
+    // this.selector.show();
```
Comment (lewis-sqa) [inline examples/angular/src/app/components/content/content.component.ts:62]: How should these be handled now?

File: examples/angular/tsconfig.json [hunk 4]
```diff
@@ -16,7 +16,7 @@
   "compilerOptions": {
     "forceConsistentCasingInFileNames": true,
     "strict": true,
-    "noImplicitOverride": true,
+    "noImplicitOverride": false,
```
Comment (lewis-sqa) [inline examples/angular/tsconfig.json:19]: What's the reason for this change?

File: examples/react/src/contexts/WalletSelectorContext.tsx [hunk 5]
```diff
@@ -27,7 +27,9 @@ interface WalletSelectorContextValue {
 const WalletSelectorContext =
   React.createContext<WalletSelectorContextValue | null>(null);
 
-export const WalletSelectorContextProvider: React.FC = ({ children }) => {
+export const WalletSelectorContextProvider: React.FC<{
+  children: React.ReactNode;
```
Comment (lewis-sqa) [inline examples/react/src/contexts/WalletSelectorContext.tsx:31]: I don't think this is needed as `React.FC` already defines this for us. You've left the `projectId` in the code too 😛

File: examples/react/src/components/Content.tsx [hunk 6]
```diff
@@ -7,6 +13,8 @@ import { useWalletSelector } from "../contexts/WalletSelectorContext";
 import SignIn from "./SignIn";
 import Form from "./Form";
 import Messages from "./Messages";
+import { Components } from "@near-wallet-selector/selector-ui/loader";
+import { WalletSelectorModal } from "@near-wallet-selector/react-selector-ui";
```
Comment (lewis-sqa) [inline examples/react/src/components/Content.tsx:17]: I'm actually a little confused about these dependencies.

When I look at [Iconic](https://ionicframework.com/react) (the main user of Stencil it seems), they have no concept of this "loader" module.

I'm also a little skeptical on whether we _actually_ need the framework specific packages if we just used Web Components directly which is supported by these frameworks and "native" Web? The reason I have this concern is it's starting to feel like the "simple modal" package installation flow is a lot more work than it should be from the perspective of a developer building a new dApp

File: oldmodal/components/WalletNotInstalled.tsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
+import React from "react";
```
Comment (lewis-sqa) [inline oldmodal/components/WalletNotInstalled.tsx:1]: We can remove this module now 👍

File: packages/core/src/index.ts [hunk 8]
```diff
@@ -55,5 +53,7 @@ export {
 
 export { FinalExecutionOutcome } from "near-api-js/lib/providers";
 
+export { WalletSelectorUIComponent } from "./lib/modal";
```
Comment (lewis-sqa) [inline packages/core/src/index.ts:56]: Don't think this should belong in core now it's "headless"

File: packages/selector-ui/src/components.d.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
+/* eslint-disable */
```
Comment (lewis-sqa) [inline packages/selector-ui/src/components.d.ts:1]: I'm wondering if there's a way to consolidate these autogenerated files we can ignore them in git. Ideally these files shouldn't be under source control

File: packages/selector-ui/src/components/wallet-selector-modal/wallet-selector-modal.e2e.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
+import { newE2EPage } from "@stencil/core/testing";
```
Comment (lewis-sqa) [inline packages/selector-ui/src/components/wallet-selector-modal/wallet-selector-modal.e2e.ts:1]: Worth removing these test files until we dedicate time to writing real tests

File: packages/selector-ui/src/components/wallet-selector/ledger-derivation-path/ledger-derivation-path.tsx [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    const wallet = await appState.selector.wallet("ledger");
+
+    if (wallet.type !== "hardware") {
+      return;
+    }
+
+    appState.ledgerIsLoading = true;
+
+    return wallet
+      .connect({ derivationPaths: [appState.ledgerDerivationPaths][0] })
```
Comment (lewis-sqa) [inline packages/selector-ui/src/components/wallet-selector/ledger-derivation-path/ledger-derivation-path.tsx:33]: I think this a typo?  It should be `derivationPaths: [appState.ledgerDerivationPaths[0]]` or simply `derivationPaths: appState.ledgerDerivationPaths`

File: packages/selector-ui/src/components/wallet-selector/readme.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
+# wallet-selector
+
+
+
+<!-- Auto Generated Below -->
+
+
+## Methods
+
+### `setRouteName(route: string) => Promise<void>`
```
Comment (lewis-sqa) [inline packages/selector-ui/src/components/wallet-selector/readme.md:10]: Not sure if we need this method?

**Review by lewis-sqa — COMMENTED** [review]
Initial pass through from me. Really nice work! 👍

File: packages/selector-ui/src/components/wallet-selector/ledger-derivation-path/ledger-derivation-path.tsx (hunk 11 above)
Comment (kujtimprenkuSQA) [reply packages/selector-ui/src/components/wallet-selector/ledger-derivation-path/ledger-derivation-path.tsx:33]: Yes it's a typo I will fix it. For now I am just adding the first element of derivationPaths but I see I did it wrongly.

File: .eslintignore (hunk 1 above)
Comment (kujtimprenkuSQA) [reply .eslintignore:6]: Yes I can remove it.

File: examples/angular/src/app/components/content/content.component.ts (hunk 3 above)
Comment (kujtimprenkuSQA) [reply examples/angular/src/app/components/content/content.component.ts:62]: We do it by getting the reference of the "wallet-selector-modal" similar like in react example, I will updatethe angular example too.

File: examples/angular/tsconfig.json (hunk 4 above)
Comment (kujtimprenkuSQA) [reply examples/angular/tsconfig.json:19]: Not sure myself, it was changed by Rares, I will ping him since I have few other questions.

File: packages/selector-ui/src/components/wallet-selector/readme.md (hunk 12 above)
Comment (kujtimprenkuSQA) [reply packages/selector-ui/src/components/wallet-selector/readme.md:10]: I was thinking it might be useful and it was an easy way to reset routeName to "WalletOptions" when closing the modal.
I will remove this and probably save routeName in store(state).

File: examples/react/src/components/Content.tsx (hunk 6 above)
Comment (kujtimprenkuSQA) [reply examples/react/src/components/Content.tsx:17]: I will investigate on this one since I am not sure about this myself.

File: packages/selector-ui/src/components/wallet-selector/readme.md (hunk 12 above)
Comment (kujtimprenkuSQA) [reply packages/selector-ui/src/components/wallet-selector/readme.md:10]: As it is now it has the same logic as before but I will try to remove that the setRouteName method.
We used to have a useEffect to check the state for the `visible `variable, not sure how to do it in stencil but I will try to implement a similar solution.

File: examples/angular/tsconfig.json (hunk 4 above)
Comment (kujtimprenkuSQA) [reply examples/angular/tsconfig.json:19]: I think this is the reason this option was set to false:
https://youtu.be/p9NaM4_CdmQ?t=271

**Comment by kujtimprenkuSQA** [comment]
Closing this PR in favor of a much simpler solution:
https://github.com/near/wallet-selector/pull/287
